src/lsm/cache_map.zig

```diff
@@ -0,0 +1,383 @@
+            // Make sure we aren't being called in regular code without another once over.
+            assert(constants.verify);
+
+            const maybe_removed = self.cache.remove(key);
+
+            if (maybe_removed) |removed| {
+                if (self.scope_is_active) {
+                    self.scope_rollback_log.appendAssumeCapacity(removed);
+                }
+            } else {
+                if (self.scope_is_active) {
+                    // TODO: Actually, does the fuzz catch this...
+                    // TODO: So if we delete from stash_2 and put to stash_1, there's a problem;
+                    //       because when we undo our scope we insert back to stash_1 :/
+                    const maybe_map_removed = self.stash_1.getKey(tombstone_from_key(key)) orelse
+                        self.stash_2.getKey(tombstone_from_key(key));
+                    if (maybe_map_removed) |map_removed| {
+                        self.scope_rollback_log.appendAssumeCapacity(map_removed);
+                    }
+                }
+            }
+
+            // We always need to try remove from the stash; since it could have a stale value.
+            _ = self.stash_1.remove(tombstone_from_key(key));
+            _ = self.stash_2.remove(tombstone_from_key(key));
+        }
+
+        /// Start a new scope. Within a scope, changes can be persisted
+        /// or discarded. At most one scope can be active at a time.
+        pub fn scope_open(self: *Self) void {
+            assert(!self.scope_is_active);
+            assert(self.scope_rollback_log.items.len == 0);
+            self.scope_is_active = true;
+        }
+
+        pub fn scope_close(self: *Self, mode: ScopeCloseMode) void {
+            assert(self.scope_is_active);
+            self.scope_is_active = false;
+
+            // We don't need to do anything to persist a scope.
+            if (mode == .persist) {
+                self.scope_rollback_log.clearRetainingCapacity();
+                return;
+            }
+
+            // The scope_rollback_log stores the operations we need to reverse the changes a scope
+            // made. They get replayed in reverse order.
+            var i: usize = self.scope_rollback_log.items.len;
+            while (i > 0) {
+                i -= 1;
+
+                const rollback_value = &self.scope_rollback_log.items[i];
+                if (tombstone(rollback_value)) {
+                    // Reverting an insert consists of a .remove call. The value in here will be a
+                    // tombstone indicating the original value didn't exist. We don't touch stash_2;
+                    // since we can never insert into it directly (only a .compact() can).
+                    const key = key_from_value(rollback_value);
+                    const removed = self.cache.remove(key) != null or
+                        self.stash_1.remove(tombstone_from_key(key));
+                    assert(removed);
```

Comment:
>  Is that the intended behavior? Iirc that wasn't the case during my last review.

Great spot - it wasn't actually intended but ended up not being a bug because of how tombstones are put there - I'll clarify it.